src/System.Management.Automation/engine/InternalCommands.cs

```diff
@@ -203,69 +251,241 @@ public object[] ArgumentList
-            Dbg.Assert(ParameterSetName == "ScriptBlockSet" || ParameterSetName == "PropertyAndMethodSet", "ParameterSetName is neither 'ScriptBlockSet' nor 'PropertyAndMethodSet'");
+            switch (ParameterSetName)
+            {
+                case "ScriptBlockSet":
+                    InitScriptBlockParameterSet();
+                    break;
 
-            if (ParameterSetName != "ScriptBlockSet") return;
+                case ForEachObjectCommand.ParallelParameterSet:
+                    InitParallelParameterSet();
+                    break;
+            }
+        }
 
-            // Win8: 176403: ScriptCmdlets sets the global WhatIf and Confirm preferences
-            // This effects the new W8 foreach-object cmdlet with -whatif and -confirm
-            // implemented. -whatif and -confirm needed only for PropertyAndMethodSet
-            // parameter set. So erring out in cases where these are used with ScriptBlockSet.
-            // Not using MshCommandRuntime, as those variables will be affected by ScriptCmdlet
-            // infrastructure (wherein ScriptCmdlet modifies the global preferences).
-            Dictionary<string, object> psBoundParameters = this.MyInvocation.BoundParameters;
-            if (psBoundParameters != null)
+        /// <summary>
+        /// Execute the processing script blocks on the current pipeline object
+        /// which is passed as it's only parameter.
+        /// </summary>
+        /// <exception cref="ParseException">Could not parse script.</exception>
+        /// <exception cref="RuntimeException">See Pipeline.Invoke.</exception>
+        /// <exception cref="ParameterBindingException">See Pipeline.Invoke.</exception>
+        protected override void ProcessRecord()
+        {
+            switch (ParameterSetName)
             {
-                SwitchParameter whatIf = false;
-                SwitchParameter confirm = false;
+                case "ScriptBlockSet":
+                    ProcessScriptBlockParameterSet();
+                    break;
 
-                object argument;
-                if (psBoundParameters.TryGetValue("whatif", out argument))
+                case "PropertyAndMethodSet":
+                    ProcessPropertyAndMethodParameterSet();
+                    break;
+
+                case ForEachObjectCommand.ParallelParameterSet:
+                    ProcessParallelParameterSet();
+                    break;
+            }
+        }
+
+        /// <summary>
+        /// Execute the end scriptblock when the pipeline is complete.
+        /// </summary>
+        /// <exception cref="ParseException">Could not parse script.</exception>
+        /// <exception cref="RuntimeException">See Pipeline.Invoke.</exception>
+        /// <exception cref="ParameterBindingException">See Pipeline.Invoke.</exception>
+        protected override void EndProcessing()
+        {
+            switch (ParameterSetName)
+            {
+                case "ScriptBlockSet":
```

Comment:
Is it possible to also make these existing parameter sets into constants?

Reply:
It is certainly possible.